drivers/counter/counter_nrfx_rtc.c

```diff
@@ -68,226 +74,428 @@ static inline const struct counter_nrfx_config *get_nrfx_config(
+	nrf_rtc_task_trigger(get_nrfx_config(dev)->rtc, NRF_RTC_TASK_CLEAR);
 
 	return 0;
 }
 
-static u32_t counter_nrfx_read(struct device *dev)
+static u32_t read(struct device *dev)
 {
-	return nrfx_rtc_counter_get(&get_nrfx_config(dev)->rtc);
+	return nrf_rtc_counter_get(get_nrfx_config(dev)->rtc);
 }
 
-static int counter_nrfx_set_alarm(struct device *dev, u8_t chan_id,
-				  const struct counter_alarm_cfg *alarm_cfg)
+/* Return true if value equals 2^n - 1 */
+static inline bool is_bit_mask(u32_t val)
 {
-	const struct counter_nrfx_config *nrfx_config = get_nrfx_config(dev);
-	const nrfx_rtc_t *rtc = &nrfx_config->rtc;
-	u32_t cc_val;
-
-	if (alarm_cfg->ticks > get_dev_data(dev)->top) {
-		return -EINVAL;
-	}
-
-	if (nrfx_config->ch_data[chan_id].callback) {
-		return -EBUSY;
-	}
+	return !(val & (val + 1));
+}
 
-	if ((alarm_cfg->flags & COUNTER_ALARM_CFG_ABSOLUTE) != 0) {
-		cc_val = alarm_cfg->ticks;
-	} else {
-		/* As RTC is 24 bit there is no risk of overflow. */
-		cc_val = alarm_cfg->ticks + nrfx_rtc_counter_get(rtc);
-		cc_val -= (cc_val > get_dev_data(dev)->top) ?
-				get_dev_data(dev)->top : 0;
+/* Function calculates distance between to values assuming that one first
+ * argument is in front and that values wrap.
+ */
+static u32_t ticks_sub(struct device *dev, u32_t val, u32_t old, u32_t top)
+{
+	if (IS_FIXED_TOP(dev)) {
+		return (val - old) & COUNTER_MAX_TOP_VALUE;
+	} else if (likely(is_bit_mask(top))) {
+		return (val - old) & top;
 	}
 
-	nrfx_config->ch_data[chan_id].callback = alarm_cfg->callback;
-	nrfx_config->ch_data[chan_id].user_data = alarm_cfg->user_data;
-
-	if ((cc_val == 0) &&
-	    (get_dev_data(dev)->top != counter_get_max_top_value(dev))) {
-		/* From Product Specification: If a CC register value is 0 when
-		 * a CLEAR task is set, this will not trigger a COMPARE event.
-		 */
-		INF("Attempt to set CC to 0, delayed to 1.");
-		cc_val++;
-	}
-	nrfx_rtc_cc_set(rtc, ID_TO_CC(chan_id), cc_val, true);
+	/* if top is not 2^n-1 */
+	return (val >= old) ? (val - old) : val + top + 1 - old;
 
-	return 0;
 }
 
-static void disable(struct device *dev, u8_t id)
+static u32_t skip_zero_on_custom_top(u32_t val, u32_t top)
 {
-	const struct counter_nrfx_config *config = get_nrfx_config(dev);
+	/* From Product Specification: If a CC register value is 0 when
+	 * a CLEAR task is set, this will not trigger a COMPARE event.
+	 */
+	if (unlikely(val == 0) && (top != COUNTER_MAX_TOP_VALUE)) {
+		val++;
+	}
 
-	nrfx_rtc_cc_disable(&config->rtc, ID_TO_CC(id));
-	config->ch_data[id].callback = NULL;
+	return val;
 }
 
-static int counter_nrfx_cancel_alarm(struct device *dev, u8_t chan_id)
+static u32_t ticks_add(struct device *dev, u32_t val1, u32_t val2, u32_t top)
 {
-	disable(dev, chan_id);
+	u32_t sum = val1 + val2;
 
-	return 0;
+	if (IS_FIXED_TOP(dev)) {
+		ARG_UNUSED(top);
+		return sum & COUNTER_MAX_TOP_VALUE;
+	}
+	if (likely(is_bit_mask(top))) {
+		sum = sum & top;
+	} else {
+		sum = sum > top ? sum - top : sum;
```

Comment:
Shouldn't it be `sum - (top + 1)` when `sum > top`?

Reply:
done